Revision 0173-Read-Generic-Network-Signal-data.md (#764)
smartdevicelink/sdl_evolution: proposals/0173-Read-Generic-Network-Signal-data.md

## Proposed solution

#### Definitions for terms used in proposal:

> * OEM Network Mapping: This file referse to key value pairs that the OEM module will use to retrieve vehicle data off its network

Reviewer: >This file referse to key

Please update to "This file refers to key..."

Author: Done

> * Vehicle Data Schema: This is a JSON object that contains vehicle data defintions for both requests and responses that can be sent using

Reviewer: >vehicle data defintions

Please update to "vehicle data definitions"

Author: Done

* All VehicleData items must be defined in the Policy Table (not just the custom/proprietary items), per recommendation by PM's Core team.
* Root-Level vehicle data type items such as `rpm` are always `mandatory`:`false` as defined by the RPC Spec for the vehicle data RPCs. If a data item is of type `Struct`, that struct can have mandatory parameters but the struct itself cannot be mandatory for the root-level item.

#### Use cases for vehicle data schema update:
This section defines SDL behavior in case _vehicle_data_items_ schema is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:

> * _vehicle_data_items_ key does not exist; do not update _vehicle_data_items_ local schema/items.

Reviewer: ```suggestion
* _vehicle_data_items_ key does not exist, do not update _vehicle_data_items_ local schema/items.
```

> * _vehicle_data_items_ exists; replace all local schema vehicle data items with new ones except the standard ones.

Reviewer: ```suggestion
* _vehicle_data_items_ exists, replace all local schema vehicle data items with new ones except the standard ones.
```

> * _vehicle_data_items_ exists but is empty; remove all local schema vehicle data items except the standard ones.

Reviewer: ```suggestion
* _vehicle_data_items_ exists but is empty, remove all local schema vehicle data items except the standard ones.
```

* _vehicle_data_items_ has item with same name as one of standard vehicle data items, ignore those data items.

> *Standard vehicle data items are the ones which are, either present in API or are provided in base schema with SDL core release. Standard vehicle data items are immutable.

Reviewer: ```suggestion
*Standard vehicle data items are the ones which are either present in API, or are provided in base schema with SDL core release. Standard vehicle data items are immutable.*
```

#### Versioning for Vehicle Data Schema
Since vehicle data schema will be downloaded as part of policy update, we should not send this data with every update when there is no change in schema. One solution could be that we include the schema in sdl_snapshot when requesting policy update, but that again does not solve the issue as snapshot gets large this way.

Solution to this is to include a schema versioning. This version information will be sent along with PTU when schema update is sent to module in PTU and SDL will persist this value in local policy DB. sdl_snapshot will contain this schema version information when requesting the policy update. Policy server would then compare the schema version received from cloud with what is available on server. Based on this comparison, vehicle data schema update would be sent in policy update response. e.g.

```
"vehicle_data_schema_version":"00.00.01",
```

#### Versioning and endpoint for OEM mapping table and to provide HMI a way to read OEM mapping table version

> OEM mapping table version would be included in PTU, which SDL will persist in local policy DB. HMI needs to be able to read this value on demand so that it can control if and when to download the OEM mapping table. HMI will utilize the _endpoint_ for OEM mapping table file to download this file using _SystemRequest_ type _OEM_SPECIFIC_. OEM mapping table file would have _endpoint_ key (_service_) as  _oem_mapping_file_.

Reviewer: ```suggestion
OEM mapping table version would be included in PTU, which SDL will persist in local policy DB. HMI needs to be able to read this value on demand so that it can control if and when to download the OEM mapping table. HMI will utilize the _endpoint_ for OEM mapping table file to download this file using _SystemRequest_ type _OEM_SPECIFIC_. OEM mapping table file would have _endpoint_ key (_service_) as _oem_mapping_file_.
```

> SDL still needs to provide an API for HMI to read this value, we can do so by adding new API to read OEM mapping file version, we can make it generic so that it can also be used in future to read some other configuration params from policy. e.g. following new API can be used to read OEM mapping table version with _identifier_ _oem_mapping_version_

Reviewer: ```suggestion
SDL still needs to provide an API for HMI to read this value. We can do so by adding a new API to read OEM mapping file version; we can make it generic so that it can also be used in the future to read some other configuration params from policy. e.g. following new API can be used to read OEM mapping table version with _identifier_ _oem_mapping_version_
```

```
<function name="GetPolicyConfigurationData" messagetype="request" scope="internal">
	<description>Request from HMI to SDL to get policy configuration data (i.e. OEM mapping table file version etc.) from Policy Table.</description>
	<param name="identifier" type="String" array="true" minsize="1" maxsize="100" maxlength="500" mandatory="true">
		<description>Id of configuration data to be received according to Policy Table i.e. oem_mapping_version etc.</description>
	</param>  
</function>
<function name="GetPolicyConfigurationData" messagetype="response">
	<param name="identifierValue" type="String" array="true" minsize="1" maxsize="100" mandatory="false">
		<description>If no message was found in PT for specified message code and for HMI current or specified language, this parameter will be omitted.</description>
	</param>
</function>
```

This API can also be extended to be used for future configuration items as well.

##### Rules for OEM mapping table and schema versioning:

> * These two variables (_oem_mapping_version_ & _vehicle_data_schema_version_) would be fed by Policy server to module during PTU. _oem_mapping_version_ would always be included in a PTU, unlike _vehicle_data_items_ and _vehicle_data_schema_version_, which would only be present if there is a change/update in vehicle data items. 

Reviewer: ```suggestion
* These two variables (_oem_mapping_version_ & _vehicle_data_schema_version_) would be fed by policy server to module during PTU. _oem_mapping_version_ would always be included in a PTU, unlike _vehicle_data_items_ and _vehicle_data_schema_version_, which would only be present if there is a change/update in vehicle data items. 
```

* SDL would need to include _vehicle_data_schema_version_ in sdl_snapshot while requesting the policy update. SDL  policy server would use this to decide whether _vehicle_data_items_ schema needs to be pushed in PTU response.
  * _vehicle_data_schema_version_ would only be included in _vehicle_data_ only if _vehicle_data_items_ schema is included.
* SDL should skip the _vehicle_data_Items_ update in case _vehicle_data_schema_version_ is not included along with it.

> * SDL would need to provide a way for HMI to read _oem_mapping_version_ so that HMI can decided whether to request OEM mapping file download using the _endpoint_ mentioned in PTU for the OEM table mapping file, using _SystemRequest_ type _OEM_SPECIFIC_

Reviewer: ```suggestion
* SDL would need to provide a way for HMI to read _oem_mapping_version_ so that HMI can decide whether or not to request OEM mapping file download using the _endpoint_ mentioned in PTU for the OEM table mapping file, using _SystemRequest_ type _OEM_SPECIFIC_
```

* Root-Level vehicle data type items such as `rpm` are always `mandatory`:`false` as defined by the RPC Spec for the vehicle data RPCs. If a data item is of type `Struct`, that struct can have mandatory parameters but the struct itself cannot be mandatory for the root-level item.

#### Use cases for vehicle data schema update:

> This section defines SDL core behavior in case `schema_items` is not present in PTU or if data in schema does not match with API/exiting schema. So if in a PTU:

Reviewer: >API/exiting schema

Should this be "API/existing schema"?

Author: Changed to RPC spec

SDL core still needs to provide an API for HMI to read this value. We can do so by adding a new API to read OEM Network Mapping file version; we can make it generic so that it can also be used in the future to read some other configuration params from policy. e.g. following new API can be used to read OEM Network Mapping table version with _identifier_ `oem_mapping_version`

```
<function name="GetPolicyConfigurationData" messagetype="request" scope="internal">
	<description>Request from HMI to SDL core to get policy configuration data (i.e. OEM Network Mapping table file version etc.) from Policy Table.</description>
	<param name="policyType" type="String" maxlength="1000" minlength="1" mandatory="true">

> 		<description>Name of the Struct from where configuration data to be received according to Policy Table types i.e. module_config etc.</description>

Reviewer: >Name of the Struct from where configuration data to be received according to Policy Table types i.e. module_config etc.

Please update to "Name of the Struct where configuration data is located in Policy Table, i.e. module_config etc."

Author: Done

	</param>  
	<param name="property" type="String" maxlength="1000" minlength="1" mandatory="true">

> 		<description>Name of the variable for which the configuration data to be received according to Policy Table types i.e. vehicle_year etc.</description>

Reviewer: >Name of the variable for which the configuration data to be received according to Policy Table types i.e. vehicle_year etc.

Please update to "Name of the property located within the policyType Struct, i.e. vehicle_year etc."

Author: Done

	</param>  
</function>
<function name="GetPolicyConfigurationData" messagetype="response">
	<param name="value" type="String" array="true" maxlength="1000" minsize="1" maxsize="100" mandatory="false">

> 		<description>Value of requested property from polocyType in PT. If no value is found in PT for specified policyType and property, this parameter will be omitted.</description>

Reviewer: >polocyType

Please update to "policyType"

Author: Done

	</param>
</function>
```

This API can also be extended to be used for future configuration items as well.

Here is the flow diagram for OEM Network Mapping table download:
![OEM%20Mapping%20table%20update%20flow](../assets/proposals/0173-Read-Generic-Network-Signal-data/OEM%20Mapping%20table%20update%20flow.png)

* HMI sends _GetPolicyConfigurationData_ request to core to read `oem_mapping_version` from policy
* Core responds with available `oem_mapping_version`

> * HMI compares the `oem_mapping_version` with local copy and dedides if OEM Network Mapping table needs to be updated

Reviewer: > dedides

Please update to "decides"

Author: N/A, outdated

* SDL core would need to include `schema_version` in sdl_snapshot while requesting the policy update. SDL server would use this to decide whether `schema_items` schema needs to be pushed in PTU response.
* `schema_version` would only be included in `vehicle_data` only if `schema_items` schema is included.
* SDL core should skip the `schema_items` update in case `schema_version` is not included along with it.
* SDL core would need to provide a way for HMI to read `oem_mapping_version` so that HMI can decide whether or not to request OEM Network Mapping file download using the _endpoint_ mentioned in PTU for the OEM table mapping file, using _SystemRequest_ `requestType` _OEM_SPECIFIC_ and  `requestSubType` _VEHICLE_DATA_MAPPING_.

Here is the flow diagram SDL server and core need to follow:
![Schema_Version_Update_flow](../assets/proposals/0173-Read-Generic-Network-Signal-data/Schema_Version_Update_flow.png)

> * `version mismatch` decision box simply checks if `schema_version` recevied in snapshot equals to `schema_version` available at server side.

Reviewer: > recevied

Please update to "received"

Author: Done

```
<function name="GetPolicyConfigurationData" messagetype="request" scope="internal">
	<description>Request from HMI to SDL Core to get policy configuration data (i.e. OEM mapping table file version etc.) from Policy Table.</description>
	<param name="policyType" type="String" maxlength="1000" minlength="1" mandatory="true">
		<description>Name of the Struct from where configuration data to be received according to Policy Table types i.e. module_config etc.</description>
	</param>  
	<param name="property" type="String" maxlength="1000" minlength="1" mandatory="true">
		<description>Name of the variable for which the configuration data to be received according to Policy Table types i.e. vehicle_year etc.</description>
	</param>  
</function>
<function name="GetPolicyConfigurationData" messagetype="response">
	<param name="value" type="String" array="true" minsize="1" maxsize="100" mandatory="false">

> 		<description>Value of requested property from polocyType in PT. If no value is found in PT for specified policyType and property, core should respond to HMI with DATA_NOT_AVAILABLE. maxlength of the value parameter is system dependent.</description>

Author: Ahh, copy/paste. Corrected it now.

* _since_ param would be used only if this vehicle data is converted to standardized type in open.

### SDL<->HMI request structure for Get/Subscribe vehicle data

In case of VehicleDataItems defined in API, SDL should send to HMI 'dataItemName : true' of item according to HMI_API.
HMI should work as is to provide response on such request. 

In case of VehicleData Items from schema, SDL should provide hierarchical structure description of vehicle data item.

> We propose to send schema vehicle data Items with following structure : 

Reviewer: ```suggestion
We propose to send schema VehicleData Items with following structure : 
```

#### Example of structure with 3 nested levels.

##### Request: 
```json
{
	"KEY_LEVEL_1" : {
		"KEY_LEVEL_2" : "true",
		"KEY2_LEVEL_2" : {
			"KEY2_LEVEL_3" : "true"
		}
	}
}
```

##### Request: 
```json
{
	"KEY_LEVEL_1" : {
		"KEY_LEVEL_2" : "value of key level 2",
		"KEY2_LEVEL_2" : {
			"KEY2_LEVEL_3" : "value of key level 3"
		}
	}
}
```
#### Backward compatibility 

Such approach is backward compatible. This also provides ability to send vehicle data items present in API and vehicle data items present in schema in one request.

#### Mix example 

Consider **gps** as API vehicle data Item, and **HEAD_LAMP_STATUS** as vehicle data item from schema: 

##### Request:
```json
{
	"gps" : "true", 
	"HEAD_LAMP_STATUS" : {
		"LOW_BEAMS_ON" : "true",
		"HIGH_BEAMS_ON" : "true",
		"AMBIENT_LIGHT_SENSOR_STATUS" : "true"
	}
}
```

##### Response: 

```json
{
	"gps" : {
		"longtitudeDegrees" : 0.0,
		"latitudeDegrees" : 0.0
	},
	"HEAD_LAMP_STATUS" : {
		"LOW_BEAMS_ON" : true,
		"HIGH_BEAMS_ON" : true,
		"AMBIENT_LIGHT_SENSOR_STATUS" : "TWILIGHT_1"
	}
}
```

#### Rules for SDL<->HMI structure for requests and responses

> - HMI should consider parameter in GVD request as name in case if it exist in list HMI_API.xml GetVehicleData of parameters.

Reviewer: ```suggestion
- HMI should consider parameter in GVD request as name in case if it exists in list HMI_API.xml GetVehicleData of parameters.
```

- HMI should consider parameter in GVD request as key if it is not in list of HMI_API.xml GetVehicleData of parameters.
- _name_ is always boolean (according to API)
- _key_ in GVD request may be either boolean or struct. Struct may contain either booleans or structs. 
- Response type for vehicle data items from schema would be _VehicleDataResult_

> - If HMI response to SDL with wrong item schema for GVD, SDL will not process this response and respond to mobile with GENERIC_ERROR

Reviewer: ```suggestion
- If HMI responds to SDL with wrong item schema for GVD, SDL will not process this response and respond to mobile with GENERIC_ERROR
```

> - If HMI send to SDL _OnVehicleData_ Notification with wrong item schema, SDL will ignore this notification.

Reviewer: ```suggestion
- If HMI sends to SDL _OnVehicleData_ Notification with wrong item schema, SDL will ignore this notification.
```
